src/v/storage/disk_log_impl.cc

```diff
@@ -339,12 +339,13 @@ ss::future<> disk_log_impl::do_compact(compaction_config cfg) {
           return !s->has_appender() && s->is_compacted_segment()
                  && !s->finished_self_compaction();
       });
-    // loop until we compact segment or reached end of segments set
+
+    // loop until we compact segment or reached last compactible segment
     for (; segit != _segs.end(); ++segit) {
         auto seg = *segit;
         if (
-          seg->has_appender() || seg->finished_self_compaction()
-          || !seg->is_compacted_segment()) {
+          segit == std::prev(_segs.end()) || seg->has_appender()
```

Comment:
not sure - is `std::prev(_segs.end())` legit if `_segs` is empty?

Reply:
there is a test in function calling this one checking if `_segs` are empty

Reply:
looks ok. you could also change iteration to stop at prev instead of end.